### PR TITLE
new python.d config files

### LINKED ISSUE #685: python.d modules configuration documentation
I suggest to add this header in all python.d/*.conf files:

```yaml
# netdata python.d.plugin configuration for ${MODULE}
#
# This file is in YaML format. Generally the format is:
#
# name: value
#
# There are 2 sections:
#  - global variables
#  - one or more JOBS
#
# JOBS allow you to collect values from multiple sources.
# Each source will have its own set of charts.
#
# JOB parameters have to be indented (example below).
#
# ----------------------------------------------------------------------
# Global Variables
# These variables set the defaults for all JOBs, however each JOB
# may define its own, overriding the defaults.
#
# update_every sets the default data collection frequency.
# If unset, the python.d.plugin default is used.
# update_every: 1
#
# priority controls the order of charts at the netdata dashboard.
# Lower numbers move the charts towards the top of the page.
# If unset, the default for python.d.plugin is used.
# priority: 60000
#
# retries sets the number of retries to be made in case of failures.
# If unset, the default for python.d.plugin is used.
# Attempts to restore the service are made once every update_every
# and only if the module has collected values in the past.
# retries: 10
#
# ----------------------------------------------------------------------
# JOBS (data collection sources)
#
# The default JOBS share the same *name*. JOBS with the same name
# are mutually exclusive. Only one of them will be allowed running at
# any time. This allows autodetection to try several alternatives and
# pick the one that works.
#
# Any number of jobs is supported.
#
# All python.d.plugin JOBS (for all its modules) support a set of
# predefined parameters. These are:
#
# job_name:
#     name: myname     # the JOB's name as it will appear at the
#                      # dashboard (by default is the job_name)
#                      # JOBs sharing a name are mutually exclusive
#     update_every: 1  # the JOB's data collection frequency
#     priority: 60000  # the JOB's order on the dashboard
#     retries: 10      # the JOB's number of restoration attempts
#
# Additionally to the above, ${MODULE} also supports the following.
#
```

where `${MODULE}` is the name of each module.

### PR DESCRIPTION
**Pull request by ktsaou**
fixes #685

Issues:

Mutually exclusion of jobs with the same name does not work. If more than one working job is found, all jobs execute in parallel.

cc: @paulfantom 

TODO:

- [x] solve the mutual exclusion issue in python.d.plugin
- [x] in charts.d.config rename the postfix, exim, phpfpm charts to append `_local`
- [x] in index.html parse properly `_local` for all python.d.plugin modules except cpufreq
- [x] update charts.d.conf so that the python modules are disabled in it
- [x] install the new configurations to users installations using the #686

### CODE REVIEW

**Comment by ktsaou** [comment]
redis:

image

mysql:

image

**Comment by paulfantom** [comment]
@ktsaou What do you mean by:
> Mutually exclusion of jobs with the same name does not work. If more than one working job is found, all jobs execute in parallel.

Can you give me configuration file which causes it? And beginning of the output from `python.d.plugin <module_name> debug` (just until it starts updates)?

I thought I fixed auto-detection with mutual exclusion yesterday.

File: conf.d/python.d/redis.conf [hunk 1]
```diff
@@ -1,14 +1,96 @@
...
-  port     : '6379'
\ No newline at end of file
+  port     : 6379
+
+localipv4:
+  name     : 'local'
+  host     : '127.0.0.1'
+  port     : 6379
+
+localipv6:
```
Comment (paulfantom) [inline conf.d/python.d/redis.conf]: I think IPv6 won't work since I am using `socket.AF_INET` to open socket and IPv6 needs `socket.AF_INET6`.

Comment (ktsaou) [inline conf.d/python.d/redis.conf]: oops! this will be a problem for several folks...

Comment (paulfantom) [inline conf.d/python.d/redis.conf]: It is only in modules based on `SocketService`. Other modules are ok with IPv6.

**Comment by ktsaou** [comment]
> Can you give me configuration file which causes it? And beginning of the output from python.d.plugin <module_name> debug (just until it starts updates)?

Here it is:

```
$ ./plugins.d/python.d.plugin debug mysql
 INFO: Using python v2
python.d INFO: reading configuration file: /etc/netdata/python.d.conf
python.d INFO: MODULES_DIR='/data/src/netdata-ktsaou.git/python.d/', CONFIG_DIR='/etc/netdata/', UPDATE_EVERY=1, ONLY_MODULES=['mysql']
python.d INFO: using MySQLdb
python.d DEBUG: mysql: loading module configuration: '/etc/netdata/python.d/mysql.conf'
python.d DEBUG: mysql: reading configuration
python.d DEBUG: mysql/tcpipv6_root: job added
python.d DEBUG: mysql/tcp: job added
python.d DEBUG: mysql/mycnf1_root: job added
python.d DEBUG: mysql/tcp_root: job added
python.d DEBUG: mysql/tcpipv6: job added
python.d DEBUG: mysql/mycnf2: job added
python.d DEBUG: mysql/socket2: job added
python.d DEBUG: mysql/socket3: job added
python.d DEBUG: mysql/socket1: job added
python.d DEBUG: mysql/tcpipv4: job added
python.d DEBUG: mysql/socket1_root: job added
python.d DEBUG: mysql/mycnf1: job added
python.d DEBUG: mysql/mycnf2_root: job added
python.d DEBUG: mysql/socket2_root: job added
python.d DEBUG: mysql/socket3_root: job added
python.d DEBUG: mysql/tcpipv4_root: job added
python.d ERROR: mysql/tcpipv6_root: is misbehaving. Reason:sequence item 2: expected string, OperationalError found
python.d ERROR: mysql/tcp: is misbehaving. Reason:sequence item 2: expected string, OperationalError found
python.d ERROR: mysql/mycnf1_root: is misbehaving. Reason:sequence item 2: expected string, OperationalError found
python.d ERROR: mysql/tcp_root: is misbehaving. Reason:sequence item 2: expected string, OperationalError found
python.d ERROR: mysql/tcpipv6: is misbehaving. Reason:sequence item 2: expected string, OperationalError found
python.d ERROR: mysql/mycnf2: is misbehaving. Reason:sequence item 2: expected string, OperationalError found
python.d ERROR: mysql/socket2: is misbehaving. Reason:sequence item 2: expected string, OperationalError found
python.d ERROR: mysql/socket3: is misbehaving. Reason:sequence item 2: expected string, OperationalError found
python.d DEBUG: mysql_socket1 : check succeeded
python.d DEBUG: mysql_socket1 changing chart name to: mysql_local
python.d DEBUG: mysql_tcpipv4 : check succeeded
python.d DEBUG: mysql_tcpipv4 changing chart name to: mysql_local
python.d ERROR: mysql_local already exists. Created with 'tcpipv4'
python.d ERROR: mysql/mycnf1: is misbehaving. Reason:sequence item 2: expected string, OperationalError found
python.d ERROR: mysql/mycnf2_root: is misbehaving. Reason:sequence item 2: expected string, OperationalError found
python.d ERROR: mysql/socket2_root: is misbehaving. Reason:sequence item 2: expected string, OperationalError found
python.d ERROR: mysql/socket3_root: is misbehaving. Reason:sequence item 2: expected string, OperationalError found
python.d DEBUG: mysql_tcpipv4_root : check succeeded
python.d DEBUG: mysql_tcpipv4_root changing chart name to: mysql_local
python.d ERROR: mysql_local already exists. Created with 'tcpipv4_root'
```

The last line is the problematic...

**Comment by ktsaou** [comment]
The configuration files are the ones in this PR

File: conf.d/python.d/redis.conf (hunk 1 above)
Comment (ktsaou) [inline conf.d/python.d/redis.conf]: I think that socket.AF_INET6 also supports IPv4, but I am not sure.

**Comment by paulfantom** [comment]
Check #689

**Comment by ktsaou** [comment]
ok, marked 1 as done.
I'll try to implement the rest tonight.

After merging this... `python.d.plugin` is production!
Thank you for your effort and work @paulfantom !

**Comment by paulfantom** [comment]
I started implementation of the second one.

**Comment by paulfantom** [comment]
check #690

**Comment by ktsaou** [comment]
done the second. thanks @paulfantom !

File: conf.d/python.d/redis.conf (hunk 1 above)
Comment (paulfantom) [inline conf.d/python.d/redis.conf]: Using ipv4 address with socket.AF_INET6 should result in ipv4 mapped to ipv6 (like `::ffff:127.0.0.1`).

I think.

Comment (ktsaou) [inline conf.d/python.d/redis.conf]: I checked what netdata daemon does: I first try ipv6. If that fails I try ipv4. If that fails too, then I return error.

Comment (paulfantom) [inline conf.d/python.d/redis.conf]: Sometime later I will do the same for `SocketService`.

**Comment by ktsaou** [comment]
I just committed a fix for No 5 (#686).

@philwhineray there is a new file `configs.signatures`. This file should be shipped with the release of netdata. It is used only by `netdata-installer.sh` to detect if the user is having a stock version of a configuration file, in which case it replaces this config file with its newer version. What do I have to do to have it in the release file?

**Comment by ktsaou** [comment]
For charts.d I did the following:

The modules that exist in python.d have to be enabled with:

```sh
module=force
```

The error messages charts.d gives reflect this. So, internally it has a list of modules that exist elsewhere and for these modules it requires this configuration to enable them.

I thought a bit about the different ways to solve it. I started writing something to parse the config file and alter it, but I didn't like it.

This solution is very clean and will work no matter how the user configuration is organized.

**Comment by ktsaou** [comment]
moved charts.d and node.d configuration to subdirectories in /etc/netdata.

The installer also moves the files it expects.

charts.d.conf will also attempt to load config files from the old location. I did that because many users may have developed their own charts.d modules and these config files will not be moved by the installer.

I think there is no need to do the same in node.d.

**Comment by ktsaou** [comment]
also updated the shebang of all shell scripts to use `/usr/bin/env bash`.

**Comment by ktsaou** [comment]
all done.

I am ready to merge it. I am going to do a few more tests before doing so.

@paulfantom once this is merged, python.d.plugin becomes mainstream. Shall we do it now or wait a bit more?

**Comment by paulfantom** [comment]
It depends if you want to have #692 fixed before `python.d.plugin` becomes mainstream or after that fact.

**Comment by ktsaou** [comment]
> It depends if you want to have #692 fixed before python.d.plugin becomes mainstream or after that fact.

I don't mind. After having this merged though, we may have a burst of issues to work on.

Have you tested it?

**Comment by paulfantom** [comment]
I have been running `python.d.plugin` on production for last week or so. I don't think I can test it more.

Also I think I can deal with burst of issues.

**Comment by ktsaou** [comment]
merge now?

**Comment by paulfantom** [comment]
Little bit scared, but do it.

**Comment by ktsaou** [comment]
merged!

**Comment by paulfantom** [comment]
I forgot about one thing (just a documentation bug). Currently we have this line:
```
Python support is experimental and implementation may change in the future
```
in `python.d/README.md` and I don't think production code can be called `experimental`.

However you don't need to change it now since tomorrow I will be editing this file to include python modules documentation.

I thought you should know.

**Comment by ktsaou** [comment]
ok

hm... just installed it. The fix you applied for mutual exclusion does not seem to work.

it says that it disabled it, but it did not (I have this for redis and mysql):

```sh
 INFO: Using python v2
python.d INFO: reading configuration file: /etc/netdata/python.d.conf
python.d INFO: MODULES_DIR='/data/src/netdata.git/python.d/', CONFIG_DIR='/etc/netdata/', UPDATE_EVERY=1, ONLY_MODULES=['redis']
python.d ERROR: redis_localipv4 No unix socket specified. Trying TCP/IP socket.
python.d ERROR: redis_localipv4 No request specified. Using: 'INFO
'
python.d ERROR: redis_localipv6 No unix socket specified. Trying TCP/IP socket.
python.d ERROR: redis_localipv6 No request specified. Using: 'INFO
'
python.d ERROR: redis_localipv6_local [Errno -9] Address family for hostname not supported
python.d ERROR: redis_localipv6_local No data received
python.d INFO: Disabled redis/local
python.d ERROR: redis/local: check() function reports failure.
python.d ERROR: redis_socket2 No request specified. Using: 'INFO
'
python.d ERROR: redis_socket2_local [Errno 2] No such file or directory
python.d ERROR: redis_socket2_local No data received
python.d INFO: Disabled redis/local
python.d ERROR: redis/local: check() function reports failure.
python.d ERROR: redis_socket3 No request specified. Using: 'INFO
'
python.d ERROR: redis_socket3_local [Errno 2] No such file or directory
python.d ERROR: redis_socket3_local No data received
python.d INFO: Disabled redis/local
python.d ERROR: redis/local: check() function reports failure.
python.d ERROR: redis_localhost No unix socket specified. Trying TCP/IP socket.
python.d ERROR: redis_localhost No request specified. Using: 'INFO
'
python.d ERROR: redis_localhost_local already exists. Stopping 'local'
python.d INFO: Disabled redis/local
CHART redis_local.operations '' Operations operations/s Statistics redis.statistics line 60000 1
DIMENSION instantaneous_ops_per_sec operations absolute 1 1
CHART redis_local.hit_rate '' 'Hit rate' percent Statistics redis.statistics line 60001 1
DIMENSION hit_rate rate absolute 1 1
CHART redis_local.memory '' 'Memory utilization' kilobytes Memory redis.memory line 60002 1
DIMENSION used_memory total absolute 1 1024
DIMENSION used_memory_lua lua absolute 1 1024
CHART redis_local.keys '' 'Database keys' keys Keys redis.keys line 60003 1
CHART redis_local.clients '' Clients clients Clients redis.clients line 60004 1
DIMENSION connected_clients connected absolute 1 1
DIMENSION blocked_clients blocked absolute 1 1
CHART redis_local.slaves '' Slaves slaves Replication redis.replication line 60005 1
DIMENSION connected_slaves connected absolute 1 1

CHART netdata.plugin_pythond_redis_local '' 'Execution time for redis_local plugin' 'milliseconds / run' python.d netdata.plugin_python area 145000 1
DIMENSION run_time 'run time' absolute 1 1

CHART redis_socket1.operations '' Operations operations/s Statistics redis.statistics line 60000 1
DIMENSION instantaneous_ops_per_sec operations absolute 1 1
CHART redis_socket1.hit_rate '' 'Hit rate' percent Statistics redis.statistics line 60001 1
DIMENSION hit_rate rate absolute 1 1
CHART redis_socket1.memory '' 'Memory utilization' kilobytes Memory redis.memory line 60002 1
DIMENSION used_memory total absolute 1 1024
DIMENSION used_memory_lua lua absolute 1 1024
CHART redis_socket1.keys '' 'Database keys' keys Keys redis.keys line 60003 1
CHART redis_socket1.clients '' Clients clients Clients redis.clients line 60004 1
DIMENSION connected_clients connected absolute 1 1
DIMENSION blocked_clients blocked absolute 1 1
CHART redis_socket1.slaves '' Slaves slaves Replication redis.replication line 60005 1
DIMENSION connected_slaves connected absolute 1 1

CHART netdata.plugin_pythond_redis_socket1 '' 'Execution time for redis_socket1 plugin' 'milliseconds / run' python.d netdata.plugin_python area 145000 1
DIMENSION run_time 'run time' absolute 1 1

BEGIN redis_local.operations 302572
SET instantaneous_ops_per_sec = 3
END
BEGIN redis_local.hit_rate 302572
SET hit_rate = 0
END
BEGIN redis_local.memory 302572
SET used_memory = 882928
SET used_memory_lua = 37888
END
BEGIN redis_local.keys 302572
END
BEGIN redis_local.clients 302572
SET connected_clients = 2
SET blocked_clients = 0
END
BEGIN redis_local.slaves 302572
SET connected_slaves = 0
END

BEGIN netdata.plugin_pythond_redis_local 302572
SET run_time = 4
END
BEGIN redis_socket1.operations 302755
SET instantaneous_ops_per_sec = 3
END
BEGIN redis_socket1.hit_rate 302755
SET hit_rate = 0
END
BEGIN redis_socket1.memory 302755
SET used_memory = 841944
SET used_memory_lua = 37888
END
BEGIN redis_socket1.keys 302755
END
BEGIN redis_socket1.clients 302755
SET connected_clients = 2
SET blocked_clients = 0
END
BEGIN redis_socket1.slaves 302755
SET connected_slaves = 0
END

BEGIN netdata.plugin_pythond_redis_socket1 302755
SET run_time = 6
END
```

**Comment by paulfantom** [comment]
Could you run redis with `debug` enabled?

**Comment by ktsaou** [comment]
> Could you run redis with debug enabled?

my last post has this.

**Comment by paulfantom** [comment]
Strange, since it doesn't show debug output. Have you run it with: `python.d.plugin redis debug 1`?

**Comment by ktsaou** [comment]
You are right. I didn't add `debug`. Here it is:

```
$ ./plugins.d/python.d.plugin redis debug 1
 INFO: Using python v2
python.d INFO: reading configuration file: /etc/netdata/python.d.conf
python.d INFO: MODULES_DIR='/data/src/netdata.git/python.d/', CONFIG_DIR='/etc/netdata/', UPDATE_EVERY=1, ONLY_MODULES=['redis']
python.d DEBUG: redis: loading module configuration: '/etc/netdata/python.d/redis.conf'
python.d DEBUG: redis: reading configuration
python.d DEBUG: redis/localipv4: job added
python.d DEBUG: redis/localipv6: job added
python.d DEBUG: redis/socket2: job added
python.d DEBUG: redis/socket3: job added
python.d DEBUG: redis/localhost: job added
python.d DEBUG: redis/socket1: job added
python.d ERROR: redis_localipv4 No unix socket specified. Trying TCP/IP socket.
python.d ERROR: redis_localipv4 No request specified. Using: 'INFO
'
python.d DEBUG: redis_localipv4_local : check succeeded
python.d DEBUG: redis_localipv4_local changing chart name to: 'redis_local'
python.d ERROR: redis_localipv6 No unix socket specified. Trying TCP/IP socket.
python.d ERROR: redis_localipv6 No request specified. Using: 'INFO
'
python.d ERROR: redis_localipv6_local [Errno -9] Address family for hostname not supported
python.d ERROR: redis_localipv6_local No data received
python.d INFO: Disabled redis/local
python.d ERROR: redis/local: check() function reports failure.
python.d ERROR: redis_socket2 No request specified. Using: 'INFO
'
python.d ERROR: redis_socket2_local [Errno 2] No such file or directory
python.d ERROR: redis_socket2_local No data received
python.d INFO: Disabled redis/local
python.d ERROR: redis/local: check() function reports failure.
python.d ERROR: redis_socket3 No request specified. Using: 'INFO
'
python.d ERROR: redis_socket3_local [Errno 2] No such file or directory
python.d ERROR: redis_socket3_local No data received
python.d INFO: Disabled redis/local
python.d ERROR: redis/local: check() function reports failure.
python.d ERROR: redis_localhost No unix socket specified. Trying TCP/IP socket.
python.d ERROR: redis_localhost No request specified. Using: 'INFO
'
python.d DEBUG: redis_localhost_local : check succeeded
python.d ERROR: redis_localhost_local already exists. Stopping 'local'
python.d INFO: Disabled redis/local
CHART redis_local.operations '' Operations operations/s Statistics redis.statistics line 60000 1
DIMENSION instantaneous_ops_per_sec operations absolute 1 1
CHART redis_local.hit_rate '' 'Hit rate' percent Statistics redis.statistics line 60001 1
DIMENSION hit_rate rate absolute 1 1
CHART redis_local.memory '' 'Memory utilization' kilobytes Memory redis.memory line 60002 1
DIMENSION used_memory total absolute 1 1024
DIMENSION used_memory_lua lua absolute 1 1024
CHART redis_local.keys '' 'Database keys' keys Keys redis.keys line 60003 1
CHART redis_local.clients '' Clients clients Clients redis.clients line 60004 1
DIMENSION connected_clients connected absolute 1 1
DIMENSION blocked_clients blocked absolute 1 1
CHART redis_local.slaves '' Slaves slaves Replication redis.replication line 60005 1
DIMENSION connected_slaves connected absolute 1 1

CHART netdata.plugin_pythond_redis_local '' 'Execution time for redis_local plugin' 'milliseconds / run' python.d netdata.plugin_python area 145000 1
DIMENSION run_time 'run time' absolute 1 1

python.d DEBUG: created charts for redis_local
CHART redis_socket1.operations '' Operations operations/s Statistics redis.statistics line 60000 1
DIMENSION instantaneous_ops_per_sec operations absolute 1 1
CHART redis_socket1.hit_rate '' 'Hit rate' percent Statistics redis.statistics line 60001 1
DIMENSION hit_rate rate absolute 1 1
CHART redis_socket1.memory '' 'Memory utilization' kilobytes Memory redis.memory line 60002 1
DIMENSION used_memory total absolute 1 1024
DIMENSION used_memory_lua lua absolute 1 1024
CHART redis_socket1.keys '' 'Database keys' keys Keys redis.keys line 60003 1
CHART redis_socket1.clients '' Clients clients Clients redis.clients line 60004 1
DIMENSION connected_clients connected absolute 1 1
DIMENSION blocked_clients blocked absolute 1 1
CHART redis_socket1.slaves '' Slaves slaves Replication redis.replication line 60005 1
DIMENSION connected_slaves connected absolute 1 1

CHART netdata.plugin_pythond_redis_socket1 '' 'Execution time for redis_socket1 plugin' 'milliseconds / run' python.d netdata.plugin_python area 145000 1
DIMENSION run_time 'run time' absolute 1 1

python.d DEBUG: created charts for redis_socket1
python.d DEBUG: redis_local will be run in 811 ms
python.d DEBUG: redis_socket1 will be run in 811 ms
python.d DEBUG: redis_local ready to run, after 812 ms (update_every: 1000 ms, latency: 0 ms
python.d DEBUG: redis_socket1 ready to run, after 813 ms (update_every: 1000 ms, latency: 2 ms
BEGIN redis_local.operations 812522
SET instantaneous_ops_per_sec = 3
END
BEGIN redis_local.hit_rate 812522
SET hit_rate = 0
END
BEGIN redis_local.memory 812522
SET used_memory = 944712
SET used_memory_lua = 37888
END
BEGIN redis_local.keys 812522
END
BEGIN redis_local.clients 812522
SET connected_clients = 3
SET blocked_clients = 0
END
BEGIN redis_local.slaves 812522
SET connected_slaves = 0
END

BEGIN netdata.plugin_pythond_redis_local 812522
SET run_time = 6
END
BEGIN redis_socket1.operations 813761
SET instantaneous_ops_per_sec = 3
END
BEGIN redis_socket1.hit_rate 813761
SET hit_rate = 0
END
BEGIN redis_socket1.memory 813761
SET used_memory = 903776
SET used_memory_lua = 37888
END
BEGIN redis_socket1.keys 813761
END
BEGIN redis_socket1.clients 813761
SET connected_clients = 3
SET blocked_clients = 0
END
BEGIN redis_socket1.slaves 813761
SET connected_slaves = 0
END
python.d DEBUG: redis_local updated in 6 ms

BEGIN netdata.plugin_pythond_redis_socket1 813761
SET run_time = 6
END
```

**Comment by ktsaou** [comment]
This is not very helpful:

```
python.d ERROR: redis/local: check() function reports failure.
```

You don't know which job...

**Comment by ktsaou** [comment]
> This is not very helpful:

Added it to #692

**Comment by ktsaou** [comment]
I am going to sleep for a few hours.
I you make a PR, I'll merge it first thing in the morning.

**Comment by paulfantom** [comment]
Fixed in #693.
I am deeply sorry, I thought I fixed it earlier.

**Comment by ktsaou** [comment]
> I am deeply sorry

no problem. Merged the fix.

**Pull request merged**